oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/OAuth2DetailedAuthenticationException.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.core;
+
+import java.util.ArrayList;
+import java.util.Collection;
+
+import org.springframework.util.Assert;
+
+/**
+ * An OAuth2 exception with additional error codes
+ *
+ * @author Josh Cummings
+ * @since 5.1
+ */
+public final class OAuth2DetailedAuthenticationException extends OAuth2AuthenticationException {
+	private Collection<OAuth2Error> details;
+
+	public OAuth2DetailedAuthenticationException(OAuth2Error error, Collection<OAuth2Error> details) {
```

Comment:
Can you see a cleaner way to do this:

```java
super(details != null && !details.isEmpty() ? details.iterator().next() : null);
```
Since I can't Assert.notEmpty(details) before calling super.

I suppose I could make it private and then have a static builder method. Thoughts?